v3/mux.go

```diff
@@ -204,20 +226,41 @@ func (m *Mux) writeLoop() {
 	}
 }
 
+func (m *Mux) pruneClosedStreams() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for id, cs := range m.closedStreams {
+		if time.Since(cs.closed) > closedStreamCleanupInterval {
+			delete(m.closedStreams, id)
+		}
+	}
+	if len(m.closedStreams) == 0 {
+		m.closedStreams = make(map[uint32]closedStream) // free memory
+	}
+}
+
 // readLoop handles the actual Reads from the Mux's net.Conn. It waits for a
 // frame to arrive, then routes it to the appropriate Stream, creating a new
 // Stream if none exists. It then waits for the frame to be fully consumed by
 // the Stream before attempting to Read again.
 func (m *Mux) readLoop() {
-	var curStream *Stream // saves a lock acquisition + map lookup in the common case
 	pr := &packetReader{
 		r:          m.conn,
 		cipher:     m.cipher,
 		packetSize: m.settings.PacketSize,
 		buf:        make([]byte, 0, m.settings.PacketSize*10),
 	}
 	frameBuf := make([]byte, m.settings.maxPayloadSize())
+
+	lastFrameCleanup := time.Now()
 	for {
+		// prune closed streams whenever enough time has passed for streams to
+		// expire
+		if time.Since(lastFrameCleanup) > closedStreamCleanupInterval {
+			m.pruneClosedStreams()
```

Comment:
why not put this in a goroutine to get it out of a hot path? You already have a mutex anyway. Only change would be needing to check that it hasn't expired.